modules/cnn_3dobj/src/cnn_feature.cpp

```diff
@@ -0,0 +1,244 @@
+#include "precomp.hpp"
+using namespace caffe;
+using std::string;
+
+namespace cv
+{
+namespace cnn_3dobj
+{
+    descriptorExtractor::descriptorExtractor(const string& device_type, int device_id)
+    {
+        if (strcmp(device_type.c_str(), "CPU") == 0 || strcmp(device_type.c_str(), "GPU") == 0)
+        {
+            if (strcmp(device_type.c_str(), "CPU") == 0)
+            {
+                caffe::Caffe::set_mode(caffe::Caffe::CPU);
+                device_info.push_back("CPU");
+                std::cout << "Using CPU" << std::endl;
+            }
+            else
+            {
+                caffe::Caffe::set_mode(caffe::Caffe::GPU);
+                caffe::Caffe::SetDevice(device_id);
+                device_info.push_back("GPU");
+                std::cout << "Using GPU" << std::endl;
+                std::cout << "Using Device_id=" << device_id << std::endl;
+            }
+            net_set = true;
+        }
+        else
+        {
+            std::cout << "Error: Device name must be 'GPU' together with an device number or 'CPU'." << std::endl;
+            net_set = false;
+        }
+    };
+
+    std::vector<string> descriptorExtractor::getDevice()
```

Comment:
split this method in 2: getDeviceType() and getDeviceId().
So you can return a single value instead than a vector of strings. It makes more sense due to the changes to setDevice (see below)